### PR TITLE
[WIP] Nested DataType compatibility in Arrow-optimized Python UDF and Pandas UDF

### PR DESCRIPTION
**Pull request by xinrong-meng**

### CODE REVIEW

File: python/pyspark/sql/pandas/serializers.py [hunk 1]
```diff
@@ -167,26 +174,23 @@ def __init__(self, timezone, safecheck, assign_cols_by_name):
...
-        # If the given column is a date type column, creates a series of datetime.date directly
-        # instead of creating datetime64[ns] as intermediate data to avoid overflow caused by
-        # datetime64[ns] type handling.
-        s = arrow_column.to_pandas(date_as_object=True)
+    def arrow_to_pandas(self, arrow_column) -> pd.Series:
+        """Convert an Arrow column to a pandas Seires."""
+        pser = arrow_column.to_pandas(date_as_object=True)
+        if type_require_conversion(arrow_column.type):
+            # We have to do per-value conversion for nested types.
+            return pser.map(lambda v: recursive_convert_inputs(v, arrow_column.type))
```
Comment (jaceklaskowski) [inline python/pyspark/sql/pandas/serializers.py:182]: "Fluent Python" 2nd Ed p. 27 says: "I used to believe that `map` and `filter` were faster than the equivalent listcomps, but Alex Martelli pointed out that's not the case". It's also said that using `listcomps` is more pythonic.

Can we rewrite the `map` to the following?

```
return [recursive_convert_inputs(v, arrow_column.type) for v in pser]
```

I'm curious if it's possible here (didn't check the types) and commenting to learn more Python 😉 

```py

```

File: python/pyspark/sql/pandas/serializers.py [hunk 2]
```diff
@@ -167,26 +174,23 @@ def __init__(self, timezone, safecheck, assign_cols_by_name):
...
-            _convert_map_items_to_dict,
-        )
-        import pyarrow
-
-        # If the given column is a date type column, creates a series of datetime.date directly
-        # instead of creating datetime64[ns] as intermediate data to avoid overflow caused by
-        # datetime64[ns] type handling.
-        s = arrow_column.to_pandas(date_as_object=True)
+    def arrow_to_pandas(self, arrow_column) -> pd.Series:
+        """Convert an Arrow column to a pandas Seires."""
```
[the remark this answers is not in the document]
Comment (xinrong-meng) [reply python/pyspark/sql/pandas/serializers.py]: Thanks!

File: python/pyspark/sql/pandas/serializers.py (hunk 1 above)
Comment (xinrong-meng) [reply python/pyspark/sql/pandas/serializers.py:182]: Thanks for sharing! We expect a pandas Series so `pser.map` would be more appropriate considering the return type.

**Comment by HyukjinKwon** [comment]
cc @haiyangsun-db FYI

**Comment by xinrong-meng** [comment]
After double thoughts, we'd better not touch Pandas UDF to preserve backward compatibility. Let me close the PR and have a new prototype.

**Pull request closed without merging**